easybuild/framework/easyconfig/easyconfig.py

```diff
@@ -677,33 +677,36 @@ def validate_iterate_opts_lists(self):
 
     def filter_hidden_deps(self):
         """
-        Filter hidden dependencies from list of (build) dependencies.
+        Replace dependencies by hidden dependencies in list of (build) dependencies, where appropriate.
         """
         dep_mod_names = [dep['full_mod_name'] for dep in self['dependencies'] + self['builddependencies']]
         build_dep_mod_names = [dep['full_mod_name'] for dep in self['builddependencies']]
 
         faulty_deps = []
-        for i, hidden_dep in enumerate(self['hiddendependencies']):
+
+        # templating must be temporarily disabled to obtain modifiable and original lists
+        # see comments in resolve_template
+        enable_templating = self.enable_templating
+        self.enable_templating = False
+        orig_hiddendeps = self['hiddendependencies']
+        self.enable_templating = enable_templating
+
+        for orig_hidden_dep, hidden_dep in zip(orig_hiddendeps, self['hiddendependencies']):
             hidden_mod_name = ActiveMNS().det_full_module_name(hidden_dep)
             visible_mod_name = ActiveMNS().det_full_module_name(hidden_dep, force_visible=True)
 
             # track whether this hidden dep is listed as a build dep
             if visible_mod_name in build_dep_mod_names or hidden_mod_name in build_dep_mod_names:
-                # templating must be temporarily disabled when updating a value in a dict;
-                # see comments in resolve_template
-                enable_templating = self.enable_templating
-                self.enable_templating = False
-                self['hiddendependencies'][i]['build_only'] = True
-                self.enable_templating = enable_templating
-
-            # filter hidden dep from list of (build)dependencies
+                orig_hidden_dep['build_only'] = True
+
+            # replace deps by hidden deps in list of (build)dependencies
             if visible_mod_name in dep_mod_names:
                 for key in ['builddependencies', 'dependencies']:
-                    self[key] = [d for d in self[key] if d['full_mod_name'] != visible_mod_name]
-                self.log.debug("Removed (build)dependency matching hidden dependency %s", hidden_dep)
+                    self[key] = [d if d['full_mod_name'] != visible_mod_name else orig_hidden_dep for d in self[key]]
+                self.log.debug("Replaced (build)dependency matching hidden dependency %s", hidden_dep)
             elif hidden_mod_name in dep_mod_names:
                 for key in ['builddependencies', 'dependencies']:
-                    self[key] = [d for d in self[key] if d['full_mod_name'] != hidden_mod_name]
+                    self[key] = [d if d['full_mod_name'] != hidden_mod_name else orig_hidden_dep for d in self[key]]
                 self.log.debug("Hidden (build)dependency %s is already marked to be installed as a hidden module",
                                hidden_dep)
```

Comment:
@bartoldeman The list comprehension with `if... else` makes this hard to read, and there's a lot of code duplication here as well.

How about something like this:

```python
# replace (build) dependencies with their equivalent hidden (build) dependency (if any) 
for key in ['builddependencies', 'dependencies']:
    deps = []
    for dep in self[key]:
        dep_mod_name = dep['full_mod_name']
        # FIXME: explain
        if visible_mod_name in dep_mod_names and dep_mod_name != visible_mod_name:
            deps.append(dep)
        # FIXME: explain
        elif hidden_mod_name in dep_mod_names and dep_mod_name != hidden_mod_name:
            deps.append(dep)
        # FIXME: explain fallback case (this is where deps are actually replaced by the equivalent hidden dep?)
        else:
            deps.append(orig_hidden_dep)
    self[key] = deps
    self.log.debug("<sensible log message goes here>")

Reply:
I have completely reorganized this code now to be iterative and replace in-place.